From d69605b9c7d31d8886449e656ed224a28c510fe2 Mon Sep 17 00:00:00 2001
From: Moritz Schulte <mo@g10code.com>
Date: Sat, 17 Jan 2009 17:19:51 +0000
Subject: [PATCH 02/52] src/pam/ChangeLog:

2009-01-17  Moritz  <moritz@gnu.org>

	* pam_poldi.c (enum opt_ids): New entry: opt_quiet.
	(opt_specs): New entry for opt_quiet.
	(pam_poldi_options_cb): Parse quiet option.
	(pam_sm_authenticate): Skip calls to conv_tell in case ctx->quiet
	is true. Thanks, Gordian.

src/pam/auth-support/ChangeLog:

2009-01-17  Moritz  <moritz@gnu.org>

	* ctx.h (struct poldi_ctx_s): New struct member: quiet.

src/pam/auth-method-localdb/ChangeLog:

2009-01-17  Moritz  <moritz@gnu.org>

	* auth-localdb.c (auth_method_localdb_auth_do): Skip calls to
	conv_tell in case ctx->quiet is true.
---
 NEWS                                       | 11 +++++++++++
 src/pam/ChangeLog                          |  8 ++++++++
 src/pam/auth-method-localdb/ChangeLog      |  5 +++++
 src/pam/auth-method-localdb/auth-localdb.c |  7 ++++---
 src/pam/auth-support/ChangeLog             |  4 ++++
 src/pam/auth-support/ctx.h                 |  5 ++++-
 src/pam/pam_poldi.c                        | 18 ++++++++++++++----
 7 files changed, 50 insertions(+), 8 deletions(-)

diff --git a/NEWS b/NEWS
index 922ee0f..551f979 100644
--- a/NEWS
+++ b/NEWS
@@ -1,5 +1,16 @@
                                                                 -*- outline -*-
 
+Changes since version 0.4.1:
+
+* New option "quiet"
+  Added a new option "quiet", which causes Poldi to skip most of the
+  PAM info messages during authentication.  Careful: the exact
+  semantics of this option might change.  Primarily this is a
+  workaround for programs like GDM, which collect these info messages
+  and put them in a dialog box with an OK-button.  When using e.g. GDM
+  with the quiet option, authentication should work without any
+  interaction.
+
 Changes since version 0.3:
 
 * Many parts have been rewritten and/or reorganized
diff --git a/src/pam/ChangeLog b/src/pam/ChangeLog
index 7a38d8b..a67c860 100644
--- a/src/pam/ChangeLog
+++ b/src/pam/ChangeLog
@@ -1,3 +1,11 @@
+2009-01-17  Moritz  <moritz@gnu.org>
+
+	* pam_poldi.c (enum opt_ids): New entry: opt_quiet.
+	(opt_specs): New entry for opt_quiet.
+	(pam_poldi_options_cb): Parse quiet option.
+	(pam_sm_authenticate): Skip calls to conv_tell in case ctx->quiet
+	is true. Thanks, Gordian.
+
 2008-12-22  Moritz  <moritz@gnu.org>
 
 	* pam_poldi.c (pam_sm_authenticate): Be more verbose in debugging
diff --git a/src/pam/auth-method-localdb/ChangeLog b/src/pam/auth-method-localdb/ChangeLog
index 8412b2a..6c20216 100644
--- a/src/pam/auth-method-localdb/ChangeLog
+++ b/src/pam/auth-method-localdb/ChangeLog
@@ -1,3 +1,8 @@
+2009-01-17  Moritz  <moritz@gnu.org>
+
+	* auth-localdb.c (auth_method_localdb_auth_do): Skip calls to
+	conv_tell in case ctx->quiet is true.
+
 2008-12-22  Moritz  <moritz@gnu.org>
 
 	* auth-localdb.c (auth_method_localdb_auth_do): Be more verbose in
diff --git a/src/pam/auth-method-localdb/auth-localdb.c b/src/pam/auth-method-localdb/auth-localdb.c
index b1b0d88..6027ee1 100644
--- a/src/pam/auth-method-localdb/auth-localdb.c
+++ b/src/pam/auth-method-localdb/auth-localdb.c
@@ -1,5 +1,5 @@
 /* auth-localdb.c - localdb authentication method for Poldi.
-   Copyright (C) 2004, 2005, 2007, 2008 g10 Code GmbH
+   Copyright (C) 2004, 2005, 2007, 2008, 2009 g10 Code GmbH
  
    This file is part of Poldi.
  
@@ -119,8 +119,9 @@ auth_method_localdb_auth_do (poldi_ctx_t ctx,
   if (ctx->debug)
     log_msg_debug (ctx->conv,
 		   _("Trying authentication as user `%s'..."), username);
-  conv_tell (ctx->conv,
-	     _("Trying authentication as user `%s'..."), username);
+  if (!ctx->quiet)
+    conv_tell (ctx->conv,
+	       _("Trying authentication as user `%s'..."), username);
 
   /* Verify (again) that the given account is associated with the
      serial number.  */
diff --git a/src/pam/auth-support/ChangeLog b/src/pam/auth-support/ChangeLog
index a7e2001..a1a64f8 100644
--- a/src/pam/auth-support/ChangeLog
+++ b/src/pam/auth-support/ChangeLog
@@ -1,3 +1,7 @@
+2009-01-17  Moritz  <moritz@gnu.org>
+
+	* ctx.h (struct poldi_ctx_s): New struct member: quiet.
+
 2008-08-17  Moritz  <moritz@gnu.org>
 
 	* ctx.h (struct poldi_ctx_s): Removed member scdaemon_socket.
diff --git a/src/pam/auth-support/ctx.h b/src/pam/auth-support/ctx.h
index 02f5e59..dfc8387 100644
--- a/src/pam/auth-support/ctx.h
+++ b/src/pam/auth-support/ctx.h
@@ -1,5 +1,5 @@
 /* ctx.h - Poldi context structure.
-   Copyright (C) 2008 g10 Code GmbH
+   Copyright (C) 2008, 2009 g10 Code GmbH
  
    This file is part of Poldi.
  
@@ -64,6 +64,9 @@ struct poldi_ctx_s
 				   should emmit debugging
 				   messages.  */
 
+  int quiet;			/* Be more quiet during PAM
+				   conversation with user. */
+
   /* Scdaemon. */
   char *scdaemon_program;	/* Path of Scdaemon program to execute.  */
   scd_context_t scd;		/* Handle for the Scdaemon access
diff --git a/src/pam/pam_poldi.c b/src/pam/pam_poldi.c
index b4883ee..6428215 100644
--- a/src/pam/pam_poldi.c
+++ b/src/pam/pam_poldi.c
@@ -1,5 +1,5 @@
 /* pam_poldi.c - PAM authentication via OpenPGP smartcards.
-   Copyright (C) 2004, 2005, 2007, 2008 g10 Code GmbH
+   Copyright (C) 2004, 2005, 2007, 2008, 2009 g10 Code GmbH
  
    This file is part of Poldi.
  
@@ -80,7 +80,8 @@ enum opt_ids
     opt_logfile,
     opt_auth_method,
     opt_debug,
-    opt_scdaemon_program
+    opt_scdaemon_program,
+    opt_quiet
   };
 
 /* Full specifications for options. */
@@ -94,6 +95,8 @@ static simpleparse_opt_spec_t opt_specs[] =
       0, SIMPLEPARSE_ARG_NONE,     0, "Enable debugging mode" },
     { opt_scdaemon_program, "scdaemon-program",
       0, SIMPLEPARSE_ARG_REQUIRED, 0, "Specify scdaemon executable to use" },
+    { opt_quiet, "quiet",
+      0, SIMPLEPARSE_ARG_NONE, 0, "Be more quiet during PAM conversation with user" },
     { 0 }
   };
 
@@ -168,6 +171,11 @@ pam_poldi_options_cb (void *cookie, simpleparse_opt_spec_t spec, const char *arg
       ctx->debug = 1;
       log_set_min_level (ctx->loghandle, LOG_LEVEL_DEBUG);
     }
+  else if (!strcmp (spec.long_opt, "quiet"))
+    {
+      /* QUIET.  */
+      ctx->quiet = 1;
+    }
 
   return gpg_error (err);
 }
@@ -480,13 +488,15 @@ pam_sm_authenticate (pam_handle_t *pam_handle,
     {
       if (ctx->debug)
 	log_msg_debug (ctx->loghandle, _("Waiting for card for user `%s'..."), pam_username);
-      conv_tell (ctx->conv, _("Waiting for card for user `%s'..."), pam_username);
+      if (!ctx->quiet)
+	conv_tell (ctx->conv, _("Waiting for card for user `%s'..."), pam_username);
     }
   else
     {
       if (ctx->debug)
 	log_msg_debug (ctx->loghandle, _("Waiting for card..."));
-      conv_tell (ctx->conv, _("Waiting for card..."));
+      if (!ctx->quiet)
+	conv_tell (ctx->conv, _("Waiting for card..."));
     }
 
   err = wait_for_card (ctx->scd, 0);
-- 
2.17.0

